feat: Add akmods module

fiftydinar opened the pull request:
Credits: @C0dePlayer

This is not ideal as it does not support custom kernels & it involves editing Containerfile.

Custom kernels is something that upstream should solve I believe, so it's not a concern we should direct to.

I believe there is no other way but to make users edit Containerfile for those files to be even pulled of.

I would like this to be through the recipe only, so I will put this as a draft until some better ideas come.

2nd part of this work:
https://github.com/ublue-os/startingpoint/pull/212

fiftydinar committed:
feat: Add kmods installer module

Credits: @C0dePlayer

This is not ideal as it does not support custom kernels & it involves editing Containerfile.

I believe there is no other way but to make users edit Containerfile for those files to be even pulled of.

I would like this to be through the recipe only, so I will put this as a draft until some better ideas come.

fiftydinar commented:
Hmm, it fails here with these errors:

<details>
  <summary>Error</summary>
  
```
=== Launching module of type: kmods-installer ===
Installing kmods

Installing: v4l2loopback winesync openrazer openrgb nct6687d gasket gcadapter_oc ryzen-smu xpadneo xone xpad-noone

Enabled rpm-md repositories: copr:copr.fedorainfracloud.org:kylegospo:rom-properties copr:copr.fedorainfracloud.org:szydell:system76 copr:copr.fedorainfracloud.org:fiftydinar:gnome-shell-openweather-toppk copr:copr.fedorainfracloud.org:kylegospo:hl2linux-selinux copr:copr.fedorainfracloud.org:kylegospo:system76-scheduler copr:copr.fedorainfracloud.org:fiftydinar:gnome-shell-extension-middleclickclose-git copr:copr.fedorainfracloud.org:fiftydinar:gnome-randr-rust rpmfusion-nonfree-updates rpmfusion-free rpmfusion-free-updates rpmfusion-nonfree copr:copr.fedorainfracloud.org:ublue-os:akmods updates fedora-cisco-openh264 fedora updates-archive

Importing rpm-md...done

rpm-md repo 'copr:copr.fedorainfracloud.org:kylegospo:rom-properties' (cached); generated: 2023-12-24T05:12:20Z solvables: 27
rpm-md repo 'copr:copr.fedorainfracloud.org:szydell:system76' (cached); generated: 2023-12-17T04:02:26Z solvables: 24
rpm-md repo 'copr:copr.fedorainfracloud.org:fiftydinar:gnome-shell-openweather-toppk' (cached); generated: 2023-11-10T12:13:13Z solvables: 2
rpm-md repo 'copr:copr.fedorainfracloud.org:kylegospo:hl2linux-selinux' (cached); generated: 2023-08-22T22:04:08Z solvables: 6
rpm-md repo 'copr:copr.fedorainfracloud.org:kylegospo:system76-scheduler' (cached); generated: 2023-12-09T09:07:11Z solvables: 4
rpm-md repo 'copr:copr.fedorainfracloud.org:fiftydinar:gnome-shell-extension-middleclickclose-git' (cached); generated: 2023-11-08T17:30:48Z solvables: 2
rpm-md repo 'copr:copr.fedorainfracloud.org:fiftydinar:gnome-randr-rust' (cached); generated: 2023-10-31T17:44:25Z solvables: 4
rpm-md repo 'rpmfusion-nonfree-updates' (cached); generated: 2023-12-22T13:50:27Z solvables: 56
rpm-md repo 'rpmfusion-free' (cached); generated: 2023-11-04T16:49:08Z solvables: 445
rpm-md repo 'rpmfusion-free-updates' (cached); generated: 2023-12-22T13:27:08Z solvables: 108
rpm-md repo 'rpmfusion-nonfree' (cached); generated: 2023-11-04T17:26:32Z solvables: 208
rpm-md repo 'copr:copr.fedorainfracloud.org:ublue-os:akmods' (cached); generated: 2023-12-09T03:59:38Z solvables: 60
rpm-md repo 'updates' (cached); generated: 2023-12-25T01:17:57Z solvables: 16809
rpm-md repo 'fedora-cisco-openh264' (cached); generated: 2023-03-14T10:57:01Z solvables: 4
rpm-md repo 'fedora' (cached); generated: 2023-11-01T00:12:39Z solvables: 70825
rpm-md repo 'updates-archive' (cached); generated: 2023-12-25T01:35:16Z solvables: 18828

Resolving dependencies...done
error: Could not depsolve transaction; 3 problems detected:
 Problem 1: conflicting requests
  - nothing provides xpadneo-kmod-common >= 0.9.5 needed by kmod-xpadneo-6.6.7-200.fc39.x86_64-0.9.5-3.20230617git5970c4c.fc39.x86_64 from @commandline
 Problem 2: conflicting requests
  - nothing provides xone-kmod-common >= 0.3 needed by kmod-xone-6.6.7-200.fc39.x86_64-0.3-3.20230517gitbbf0dcc.fc39.x86_64 from @commandline
 Problem 3: conflicting requests
  - package kernel-devel-matched-6.6.7-200.fc39.x86_64 from updates-archive requires kernel-devel = 6.6.7-200.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.6.8-200.fc39.x86_64 from updates requires kernel-core = 6.6.8-200.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.6.8-200.fc39.x86_64 from updates-archive requires kernel-core = 6.6.8-200.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.5.6-300.fc39.x86_64 from fedora requires kernel-core = 6.5.6-300.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.5.10-300.fc39.x86_64 from updates-archive requires kernel-core = 6.5.10-300.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.5.11-300.fc39.x86_64 from updates-archive requires kernel-core = 6.5.11-300.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.5.12-300.fc39.x86_64 from updates-archive requires kernel-core = 6.5.12-300.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.5.9-300.fc39.x86_64 from updates-archive requires kernel-core = 6.5.9-300.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.6.2-201.fc39.x86_64 from updates-archive requires kernel-core = 6.6.2-201.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.6.3-200.fc39.x86_64 from updates-archive requires kernel-core = 6.6.3-200.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.6.4-200.fc39.x86_64 from updates-archive requires kernel-core = 6.6.4-200.fc39, but none of the providers can be installed
  - package kernel-devel-matched-6.6.6-200.fc39.x86_64 from updates-archive requires kernel-core = 6.6.6-200.fc39, but none of the providers can be installed
  - cannot install both kernel-devel-6.6.7-200.fc39.x86_64 from updates-archive and kernel-devel-6.6.8-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.6.8-200.fc39.x86_64 from updates and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.5.6-300.fc39.x86_64 from fedora and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.5.10-300.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.5.11-300.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.5.12-300.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.5.9-300.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.6.2-201.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.6.3-200.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.6.4-200.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.6.6-200.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
  - cannot install both kernel-core-6.6.8-200.fc39.x86_64 from updates-archive and kernel-core-6.6.7-200.fc39.x86_64 from @System
error building at STEP "RUN chmod +x /tmp/build.sh && /tmp/build.sh &&     rm -rf /tmp/* /var/* && ostree container commit": error while running runtime: exit status 1
time="2023-12-25T20:44:24Z" level=error msg="exit status 1"
```

</details>

xynydev commented on modules/kmods-installer/README.md:
```diff
@@ -0,0 +1,31 @@
+# `kmods-installer` Module for Startingpoint
```
My understanding is that this only installs from https://github.com/ublue-os/akmods, so maybe just `akmods` (or maybe `akmods-installer`) would be a better name, since for https://github.com/ublue-os/bling the module name is just `bling`. Usually adding in `-installer`, `-runner`, `-configurator` or something like that is redundant, too.

fiftydinar replied:
I agree, I will change this

xynydev commented on modules/kmods-installer/README.md:
```diff
@@ -0,0 +1,31 @@
+## Notes
+
+It should be used for Fedora 39 & above only.
+For Nvidia, use Universal Blue base Nvidia images + akmod builds with regular tag.
```
> For Nvidia, use Universal Blue base Nvidia images + akmod builds with regular tag.

This line is a bit confusing, does this mean that Nvidia builds shouldn't use the module at all?  
This section could also be made to stand out more, maybe just by adding it as a warning right before the example config, reading something like:  
_Warning: only use this module with non-Nvidia builds based on Fedora 39 or older. <Explanation for why?> Only Universal Blue images are officially supported. Custom kernels are not supported, because..._

fiftydinar replied:
This means that Nvidia users should use -nvidia builds as a base because it already contains needed nvidia kmod. 
So they can add the modules they want normally like the users who base from -main images.

Maybe some Nvidia users would try to use main as a base + add Nvidia akmod, because they think that way is "cleaner" deluding themselfs. But that is not necessary as -nvidia builds already cover them.

I will try to rename this to be clearer or maybe even remove it completely as it can be seen as confusing.

fiftydinar replied:
Addressed

xynydev commented:
> Hmm, it fails here with these errors

The akmods repo example installs the akmods with the commands below. Maybe the wildcard is required?
```
RUN rpm-ostree install /tmp/rpms/ublue-os/ublue-os-akmods*.rpm
RUN rpm-ostree install /tmp/rpms/kmods/kmod-v4l2loopback*.rpm
```

fiftydinar committed:
Rename kmods installer to akmods

xynydev commented:
I was just looking at if there'd be a way to install the akmods directly using the module without Containerfile changes, and there _is_ [a COPR](https://copr.fedorainfracloud.org/coprs/ublue-os/akmods/) but it doesn't have everything and I'm not sure if it'd _just work_.

With a compiler-based approach this would be easier. I guess I should work on that sometime...

fiftydinar committed:
Merge remote-tracking branch 'upstream/main' into kmods-installer

fiftydinar committed:
Update README

Address: 

https://github.com/ublue-os/startingpoint/pull/212#issuecomment-1870156327

https://github.com/ublue-os/bling/pull/89#discussion_r1436890002

fiftydinar committed:
Fix README typo

fiftydinar commented:
@xynydev I have great news.

Module actually works!
Those errors I had before seems to be a typical upstream error (from akmods repo).

Here's proof it works:
https://github.com/fiftydinar/gidro-os/actions/runs/7473504118/job/20337794147#step:15:647

I think this is good to merge, I don't see what more of a changes is needed.

If you catch something that needs update, please let me know.

Edit: I remembered. We need to solve ${IMAGE_MAJOR_VERSION} tag in Containerfile for pulling akmods files instead of fixed Fedora version.

However, adding IMAGE_MAJOR_VERSION as an argument only works when building an image offline. Github Action fails.

So I think it's good as it is (not yet, see below).

It's a shame that akmods repo doesn't have `latest` tag available though. So you need to edit this manually whenever new Fedora gets a release.
I filled an issue on this: https://github.com/ublue-os/akmods/issues/109

fiftydinar committed:
Remove non-needed space for yml in README

fiftydinar committed:
Add support for Surface & Asus images

fiftydinar commented:
I added support for Surface & Asus images as I noticed they don't install kernel-devel-matched package.

It auto-detects if the base image is right before installing kmods.

However, it seems that BASE_IMAGE variable is not recognized in cloud for some reason, hence builds failing.
I'll have to investigate this further. Help is also appreciated.

fiftydinar committed:
Clarify tagged base image warning better

fiftydinar committed:
Clarify tagged base image warning better pt.2

fiftydinar committed:
There is no need to fetch main-nvidia build for now

fiftydinar commented:
Here's proof that BASE_IMAGE variable works in terminal (but still not working in cloud):
![Screenshot from 2024-01-12 14-00-35](https://github.com/ublue-os/bling/assets/65243233/5e2c9f13-c3b6-48ee-b4e4-7149b8a721ae)

When I put something simple in module like
`echo "${BASE_IMAGE}"`

it echoes base image normally

fiftydinar committed:
Use simpler =~ for conditioning instead of grep & sed

This finally fixes akmods module

fiftydinar committed:
Install kernel-devel-matched for all builds

fiftydinar commented:
Installing kernel-devel-matched actually works for all images & it's the more proper way of installing akmods.
I tested surface, asus & main image as a base & module works wonderfully.

This time, it's really ready to merge.

fiftydinar committed:
Assure that Surface installs their version of kernel-devel-matched

fiftydinar committed:
Mention that framework images can be used as a base

fiftydinar committed:
Delete duplicate warning message

fiftydinar committed:
Remove non-needed explanation on why only Universal Blue builds are supported

fiftydinar committed:
Clarify 1st warning better

fiftydinar committed:
Clarify `main` akmods compatibility better

This would avoid editing README if some other compatible image gets announced or discontinued.

xynydev committed:
docs(akmods): grammar fixes

xynydev commented on modules/akmods/akmods.sh:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+set -oue pipefail
+
+BASED_IMAGE=$(echo "${BASE_IMAGE}")
```
What's the point of this? The `BASE_IMAGE`and `BASED_IMAGE` variable are identical, no?

fiftydinar replied:
You're right, there is no need for this.

I can correct this in another PR.

xynydev committed:
docs: add link to akmod tag matrix
